api/src/utils/calculate-field-depth.ts

```diff
@@ -0,0 +1,79 @@
+import { isPlainObject, isArray } from 'lodash';
+
+/**
+ * Calculates the depth of a given JSON structure, not counting any _ prefixed properties
+ *
+ * Used to calculate the field depth in a filter or deep query structure
+ *
+ * @example
+ *
+ * ```js
+ * const deep = {
+ * 	translations: {
+ * 		_filter: {
+ * 			_and: [
+ * 				{
+ * 					language_id: {
+ * 						name: {
+ * 							_eq: 'English'
+ * 						}
+ * 					}
+ * 				},
+ * 				{
+ * 					status: {
+ * 						_eq: 'Published'
+ * 					}
+ * 				}
+ * 			]
+ * 		}
+ * 	}
+ * };
+ *
+ * const result = calculateFieldDepth(deep); // => 3
+ * ```
+ */
+export function calculateFieldDepth(obj?: Record<string, any> | null): number {
+	if (!obj) {
+		return 0;
+	}
+
+	let depth = 0;
+
+	const keys = Object.keys(obj);
+
+	for (const key of keys) {
+		const nestedValue = obj[key];
+
+		if (key === '_sort' && nestedValue) {
```

Comment:
I'm not sure what should be in the `validateAST`... 🤔
Currently there's only the validation of relational field depths in `api/src/utils/get-ast-from-query.ts`. 

As for the `_sort` case, it's a tricky one as the `_sort` could be in any depth. The depth above the `_sort` has to be added to form the total... 

Can we add a second parameter to `calculateFieldDepth()` with a boolean on whether it's to calculate a `deep` filter? 
Example: `calculateFieldDepth(obj, isDeepFilter)`

Reply:
> Currently there's only the validation of relational field depths in api/src/utils/get-ast-from-query.ts.

I was just wondering if it would make more sense to keep the generation of the AST and the validation of the outcome separate 🤔 It feels a little cleaner to me to have a dedicated step for validation, rather than include it in a generator function, so it'll become:

```
(incoming request)
-> sanitize query
-> validate query
-> generate AST
-> validate AST
-> run AST
```

> As for the _sort case, it's a tricky one as the _sort could be in any depth. The depth above the _sort has to be added to form the total...

Ah yeah good one 🤔 In that case, a secondary parameter to enable it should work. Just trying to keep this as clean as possible. Maybe we should consider splitting it up into 3 different functions: One to calculate field depth from an object, one to calculate the depth of fields in an array of dot-notated strings, and one loop over all keys and select between the two recursively 🤔

Reply:
Refactored to validate relational depth under validateQuery.
Also added another parameter to `calculateFieldDepth()`. 

I'm unsure about the splitting up as a user's input value might contain dot-notated strings. Eg: '{email: {_contains: "a.b.c@example.com"} }'

Reply:
> I'm unsure about the splitting up as a user's input value might contain dot-notated strings.

Yeah! That's why we'd have a single recursively looping function that chooses between nested object vs dot notation based on the key name (so you can check against `_sort` and `_fields` 🙂

Reply:
@rijkvanzanten Instead of using a boolean, it is process dot notation style based on an array of keys.
I'm not quite sure of how it should be split up, please assist to refactor as necessary.
Thank you~ 🙏🏼

Reply:
Looking good now!